Crash when opening a certain project (directory)

### Check for Existing Issues

- [x] Completed

### Bug Description

Zed IDE crashes when opening certain projects or when launching it with the last opened project. This results in a recurring crash on startup since the last project is automatically loaded. To circumvent this issue, I must start Zed in a different directory that is not associated with the problematic project, e.g., `zed /some/another/path`.

### Environment

Zed: v0.157.5 (Zed)
OS: Linux Wayland bluefin 40
Memory: 31 GiB
Architecture: x86_64
GPU: Intel(R) Xe Graphics (TGL GT2) || Intel open-source Mesa driver || Mesa 24.1.6

### Log File

The relevant error that likely leads to the crash is as follows:

<details>
<summary>Error that i suppose crashes the program</summary>

```
2024-10-20T18:28:20.58325496+04:00 [ERROR] {
  "thread": "main",
  "payload": "`at` split index (is 4) should be <= len (is 1)",
  "location_data": {
    "file": "/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/alloc/src/vec/mod.rs",
    "line": 2319
  },
  "backtrace": [
    "core::panicking::panic_fmt",
    "alloc::vec::Vec<T,A>::split_off::assert_failed",
    "workspace::pane::Pane::render_tab_bar",
    "<workspace::pane::Pane as gpui::element::Render>::render",
    "gpui::view::any_view::render",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint::{{closure}}",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}::{{closure}}",
    "gpui::elements::div::Interactivity::prepaint::{{closure}}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<E> as gpui::element::ElementObject>::prepaint",
    "gpui::element::AnyElement::prepaint",
    "gpui::window::WindowContext::draw",
    "gpui::window::WindowContext::dispatch_event",
    "gpui::window::Window::new::{{closure}}",
    "gpui::platform::linux::wayland::window::WaylandWindowStatePtr::handle_input",
    "<gpui::platform::linux::wayland::client::WaylandClientStatePtr as wayland_client::event_queue::Dispatch<wayland_client::protocol::wl_keyboard::WlKeyboard,()>>::event",
    "wayland_client::event_queue::queue_callback",
    "<core::cell::RefCell<calloop::sources::DispatcherInner<S,F>> as calloop::sources::EventDispatcher<Data>>::process_events",
    "<gpui::platform::linux::wayland::client::WaylandClient as gpui::platform::linux::platform::LinuxClient>::run",
    "gpui::platform::linux::platform::<impl gpui::platform::Platform for P>::run",
    "gpui::app::App::run",
    "zed::main",
    "std::sys::backtrace::__rust_begin_short_backtrace",
    "std::rt::lang_start::{{closure}}",
    "std::rt::lang_start_internal",
    "main",
    "__libc_start_call_main",
    "__libc_start_main@@GLIBC_2.34",
    "_start"
  ],
  "app_version": "0.157.5",
  "release_channel": "Zed",
  "os_name": "Linux Wayland",
  "os_version": "bluefin 40",
  "architecture": "x86_64",
  "panicked_on": 1729434500583,
  "system_id": "3acdd293-70c0-4183-a519-6b76539d16d4",
  "installation_id": "4bc8b480-c404-4042-9edf-81061d7796e9",
  "session_id": "1b8ac22c-5a9b-403a-9dd8-5050ad674ecb"
}
}
```

</details>

Complete log including the error: [Zed.log](https://github.com/user-attachments/files/17450702/Zed.log)

--- a/crates/workspace/src/pane.rs
+++ b/crates/workspace/src/pane.rs
@@ -1593,8 +1593,13 @@ impl Pane {
                             Ok(0) => {}
                             Ok(1) => {
                                 // Don't save this file
-                                pane.update(cx, |_, cx| item.discarded(project, cx))
-                                    .log_err();
+                                pane.update(cx, |pane, cx| {
+                                    if pane.is_tab_pinned(item_ix) && !item.can_save(cx) {
+                                        pane.pinned_tab_count -= 1;
+                                    }
+                                    item.discarded(project, cx)
+                                })
+                                .log_err();
                                 return Ok(true);
                             }
                             _ => return Ok(false), // Cancel

